### PR TITLE
Add "status" request helper

### PR DESCRIPTION
**Pull request by dashie**
**Technical implementation details**

I want to propose this new helper. The impact is very little and it is simple to be used, but can be very useful in many cases.
The idea is to be able to set the response status inside the template.

Imagine a route like this:

    test/:id

and a data bucket with name `testdata` and the following content:

```
{
  "data": {
    "001": {
      "id": "001",
      "name": "Item 1"
    },
    "002": {
      "id": "002",
      "name": "Item 2"
    }
  }
}
```

now using a template like this:

```handlebar
{{setVar 'selection' (concat 'data.' (urlParam 'id'))}}
{{#if (dataRaw 'testdata' @selection true)}}
{{data 'testdata' @selection}}
{{else}}
{{status 404}}
{
  "error": "NOT_EXISTS"
}
{{/if}}
```

we have a routes where we can return 200 with data if the data exists in the bucket

    GET /test/001

or 404 if the data does not exist.

    GET /test/003

To be able to check the path selection I've also added an optional flag on `dataRaw` helper, that return an empty string `""` when the selection path does not match any property.

*If you accept this proposal I can finish it with tests and docs :P*

**Checklist**

- [x] data migration added (@mockoon/commons)
- [x] data migration automated tests added (@mockoon/commons)
- [x] CLI automated tests added (@mockoon/cli)
- [x] desktop automated tests added (@mockoon/desktop)

### CODE REVIEW

**Comment by 255kb** [comment]
Thank you for opening this PR. 
I have mixed feelings regarding this feature. As stated in #893, CRUD endpoints will have this kind of behavior. They will be linked to a data bucket and return a 404 when something is missing. The implementation will be similar to what json-server does. 
Introducing this `status` helper is clever, but I have the feeling that it's a bit redundant with the CRUD routes (future) implementation. 
Anyway, the feature is kind of hidden in a helper and it's not adding more complexity to the UI or anything like this. 
So, I guess it can be added 😄.

I will review and probably add some remarks. I saw that the linter is failing. Maybe correct this first so we can run all the tests.

File: packages/commons-server/src/libs/server/server.ts [hunk 1]
```diff
@@ -531,6 +531,13 @@ export class MockoonServer extends (EventEmitter as new () => TypedEmitter<Serve
         );
       }
 
+      if (request.locals) {
+        const statusCode = request.locals.get('statusCode');
+        if (statusCode) {
+          response.status(statusCode);
+        }
+      }
+
```
Comment (255kb) [inline packages/commons-server/src/libs/server/server.ts:540]: This should be replicated in the `sendFile` method too (line 627-628), as the templating system is supported in files.

File: packages/commons-server/src/libs/templating-helpers/data-helpers.ts [hunk 2]
```diff
@@ -83,7 +83,11 @@ export const DataHelpers = function (
 
         // ensure a value was found at path
         const foundValue = objectGet(value, path);
-        value = foundValue !== undefined ? foundValue : value;
+        if (parameters.length > 1 && parameters[2]) {
+          value = foundValue !== undefined ? foundValue : '';
+        } else {
+          value = foundValue !== undefined ? foundValue : value;
+        }
```
Comment (255kb) [inline packages/commons-server/src/libs/templating-helpers/data-helpers.ts:90]: Could you please put this in a separated PR? I prefer keeping each PR focused on one topic. 
Also, this is clearly a discrepancy between how data/dataRaw and body/bodyRaw helpers are working. So, the change is needed but I would rather think this thoroughly in a separated task. (I can also handle it, in which case you can remove it altogether)

File: packages/commons-server/src/libs/templating-helpers/request-helpers.ts [hunk 3]
```diff
@@ -24,6 +24,17 @@ export const RequestHelpers = function (
...
+    status: function (status: any) {
+      const intCode = parseInt(status, 10);
+      if (isNaN(intCode)) {
+        return;
+      }
+      if (!request.locals) {
+        request.locals = new Map<string, any>();
+      }
+      request.locals.set('statusCode', intCode);
+    },
```
Comment (255kb) [inline packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: As this is the first helper dealing with the response, I would prefer to put this in the helpers.ts file or in a new response-helpers.ts file.

**Review by 255kb — CHANGES_REQUESTED** [review]

File: packages/commons-server/src/libs/server/server.ts (hunk 1 above)
Comment (dashie) [reply packages/commons-server/src/libs/server/server.ts:540]: Ok, no prob, I'm going to add it on sendFile too

File: packages/commons-server/src/libs/templating-helpers/data-helpers.ts (hunk 2 above)
Comment (dashie) [reply packages/commons-server/src/libs/templating-helpers/data-helpers.ts:90]: So do you want I add this behavior on:
- data/dataRaw
- body/bodyRaw
and that I create a specific new PR for this change?

File: packages/commons-server/src/libs/templating-helpers/request-helpers.ts (hunk 3 above)
Comment (dashie) [reply packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: We need to move also the test in a separate file?

Comment (255kb) [reply packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: yes I tried to matche the lib/test filenames

File: packages/commons-server/src/libs/templating-helpers/data-helpers.ts (hunk 2 above)
Comment (255kb) [reply packages/commons-server/src/libs/templating-helpers/data-helpers.ts:90]: Don't worry, I will do it.
But anyway, PR may be merged but the next release will only occur around January.

Comment (dashie) [reply packages/commons-server/src/libs/templating-helpers/data-helpers.ts:90]: tnx, keep us updated if it is possible

File: packages/commons-server/src/libs/templating-helpers/request-helpers.ts (hunk 3 above)
Comment (dashie) [reply packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: so I wait for your change

Comment (255kb) [reply packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: Not sure I understand what you mean. Do you want me to move the code in a new file? Let me know and I will do it.

Comment (dashie) [reply packages/commons-server/src/libs/templating-helpers/request-helpers.ts:37]: Yes, if it is needed and you can do it is great, tnx

**Comment by 255kb** [comment]
I realized that this helper is not compatible with the data buckets as they are generated during server start, or during the first call. 
I will finalize this PR, and add some documentation. 
It will be part of the release in a few days.

**Pull request merged**